File: app/src/main/java/com/hhy/game/snowman/game/Animator.java
package com.hhy.game.snowman.game;

import android.graphics.Bitmap;
import android.graphics.Rect;
import java.util.ArrayList;

class Animator {
    private Bitmap bitmapWithFrames;
    private int updateTimeMillis;
    private ArrayList<Rect> frames = new ArrayList<>();
    private int currentFrame;
    private double timeForCurrentFrameMillis;
    private double timeForOneFrameMillis;
    private AnimatorThread animatorThread;
    private volatile boolean isRunning;
    private int frameWidth;
    private int frameHeight;
    private Player player;

    Animator(Player player, Bitmap bitmapWithFrames, int numberOfFramesHorizontally,
                    int numberOfFramesVertically, int updateTimeMillis,
                    double timeForOneFrameMillis) {
        this.player = player;
        this.bitmapWithFrames = bitmapWithFrames;
        this.updateTimeMillis = updateTimeMillis;
        this.timeForOneFrameMillis = timeForOneFrameMillis;
        splitBitmapIntoFrames(bitmapWithFrames, numberOfFramesHorizontally, numberOfFramesVertically, frames);
    }

    void startAnimation() {
        isRunning = true;
        animatorThread = new AnimatorThread();
        animatorThread.setDaemon(true);
        animatorThread.start();
    }

    void stopAnimation() {
        isRunning = false;
        boolean flag = true;

        while (flag) {
            try {
                animatorThread.join();
                flag = false;
            } catch (InterruptedException ignored) {}
        }
    }

    boolean isRunning() {
        return isRunning;
    }

    private void splitBitmapIntoFrames(Bitmap source, int numberOfFramesHorizontally,
                                       int numberOfFramesVertically, ArrayList<Rect> listDestination) {
        frameWidth = source.getWidth() / numberOfFramesHorizontally;
        frameHeight = source.getHeight() / numberOfFramesVertically;

        for (int i = 0; i < numberOfFramesVertically; i++)
            for (int j = 0; j < numberOfFramesHorizontally; j++) {
                listDestination.add(new Rect(j * frameWidth, i * frameHeight,
                        j * frameWidth + frameWidth, i * frameHeight + frameHeight));
            }
    }

    private class AnimatorThread extends Thread {
        private Bitmap newBitmap;

        @Override
        public void run() {
            while (isRunning) {
                timeForCurrentFrameMillis += updateTimeMillis;
                if (timeForCurrentFrameMillis >= timeForOneFrameMillis) {
                    currentFrame = (currentFrame + 1) % frames.size();
                    timeForCurrentFrameMillis = timeForCurrentFrameMillis - timeForOneFrameMillis;
                }

                newBitmap = Bitmap.createBitmap(bitmapWithFrames, frames.get(currentFrame).left,
                        frames.get(currentFrame).top, frameWidth,
                        frameHeight);

                player.setBitmap(newBitmap);

                try {
                    sleep(updateTimeMillis);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        }
    }
}

File: app/src/main/java/com/hhy/game/snowman/game/Animator.kt
package com.hhy.game.snowman.game

import android.graphics.Bitmap
import android.graphics.Rect
import java.util.*

internal class Animator(private val player: Player, private val bitmapWithFrames: Bitmap, numberOfFramesHorizontally: Int,
    numberOfFramesVertically: Int, private val updateTimeMillis: Int, private val timeForOneFrameMillis: Double) {
    private val frames = ArrayList<Rect>()
    private var currentFrame = 0
    private var timeForCurrentFrameMillis = 0.0
    private lateinit var animatorThread: AnimatorThread

    @Volatile
    var isRunning = false
        private set
    private var frameWidth = 0
    private var frameHeight = 0

    fun startAnimation() {
        isRunning = true
        animatorThread = AnimatorThread()
        animatorThread.isDaemon = true
        animatorThread.start()
    }

    fun stopAnimation() {
        isRunning = false
        var flag = true
        while (flag) {
            try {
                animatorThread.join()
                flag = false
            } catch (ignored: InterruptedException) {
            }
        }
    }

    private fun splitBitmapIntoFrames(source: Bitmap, numberOfFramesHorizontally: Int, numberOfFramesVertically: Int,
                                      listDestination: ArrayList<Rect>) {
        frameWidth = source.width / numberOfFramesHorizontally
        frameHeight = source.height / numberOfFramesVertically
        for (i in 0 until numberOfFramesVertically) for (j in 0 until numberOfFramesHorizontally) {
            listDestination.add(
                Rect(
                    j * frameWidth, i * frameHeight,
                    j * frameWidth + frameWidth, i * frameHeight + frameHeight
                )
            )
        }
    }

    private inner class AnimatorThread : Thread() {
        private lateinit var newBitmap: Bitmap
        override fun run() {
            while (isRunning) {
                timeForCurrentFrameMillis += updateTimeMillis.toDouble()
                if (timeForCurrentFrameMillis >= timeForOneFrameMillis) {
                    currentFrame = (currentFrame + 1) % frames.size
                    timeForCurrentFrameMillis -= timeForOneFrameMillis
                }
                newBitmap = Bitmap.createBitmap(
                    bitmapWithFrames, frames[currentFrame].left,
                    frames[currentFrame].top, frameWidth,
                    frameHeight
                )
                player.setBitmap(newBitmap)

                sleep(updateTimeMillis.toLong())
            }
        }
    }

    init {
        splitBitmapIntoFrames(bitmapWithFrames, numberOfFramesHorizontally, numberOfFramesVertically, frames)
    }
}